src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs

```diff
@@ -397,5 +399,42 @@ public bool DetermineExtensionDataProperty(Dictionary<string, JsonPropertyInfo>
 
             return converter;
         }
+
+        public static void ValidateType(Type type, Type? parentClassType, PropertyInfo? propertyInfo)
+        {
+            if (IsInvalidForSerialization(type))
+            {
+                ThrowHelper.ThrowInvalidOperationException_CannotSerializeInvalidType(type, parentClassType, propertyInfo);
+            }
+        }
+
+        public static bool IsInvalidForSerialization(Type type)
+        {
+            return type.IsPointer || IsByRefLike(type) || type.ContainsGenericParameters;
+        }
+
+        public static bool IsByRefLike(Type type)
+        {
+#if BUILDING_INBOX_LIBRARY
+            return type.IsByRefLike;
+#else
+            if (!type.IsValueType)
+            {
+                return false;
+            }
+
+            object[] attributes = type.GetCustomAttributes(inherit: false);
+
+            for (int i = 0; i < attributes.Length; i++)
+            {
+                if (attributes[i].GetType().FullName == "System.Runtime.CompilerServices.IsByRefLikeAttribute")
+                {
+                    return true;
+                }
+            }
+
+            return false;
```

Comment:
Does the result of this end up getting cached at a higher-level?  Just want to make sure we're not incurring this reflection on every operation.

Reply:
Yes. This information is fetched once per type to (de)serialize (during warm up), and won't be fetched again on subsequent (de)serialization operations.